### PR TITLE
Add descriptive error messages to unwrap()s

### PR DESCRIPTION
**Pull request by cmannett85**
## Reason for This PR
#593 

## Description of Changes
Added these new strings to the codebase:
- "Cannot start HTTP server"
- "Cannot update send VMM fd"
- "VMM disconnected"
- "Cannot convert from UTF-8"
- "No field key"
- "Field value not a string"
- "Time conversion overflow"
- "Failed to clone API event FD"
- "EventManager exiting"
- "Poisoned subscriber lock"
- "Poisoned VMM lock"
- "Unexpected VirtioDevice type"
- "IRQ chip not set"
- "vCPU already started"
- "Poisoned legacy serial lock"
- "Failed to parse version"
- "Failed to parse ident"
- "Failed to parse array length"

I did not update `unwrap()`s in tests or example code, change existing `expect()`s, or update `unwrap()`s where the comments/context stated it was safe.  If this _is_ desirable, please say!

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance
By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist
- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

**Comment by iulianbarbu** [comment]
Hey, @cmannett85 ! Thanks for the PR.

I'll look on the changes in depth soon. A high level comment would be that this PR has to provide descriptive error messages which might help users in solving Firecracker related problems by themselves. Some of these `expect`s from your changes address also debugging messages, useful for developers. These messages contain implementation details which IMO do not bring value to end users. However, this does not mean that we shouldn't have messages which tell users that an error occurred, due to a reason independent to them. We might have to decide upon these.

**Comment by cmannett85** [comment]
@iulianbarbu that's fine, I'm new to Firecracker so I picked up this issue as it forces me to look around a nice varied selection of the code base, but that obviously means that I don't understand some of the context surrounding the `expect()`s - I'm happy to change any of it as it's a great way to learn.

**Comment by cmannett85** [comment]
In aarch64 it is failing due to this:

**Comment by dianpopa** [comment]
> In aarch64 it is failing due to this:
> 
> ```
> error[E0716]: temporary value dropped while borrowed
>    --> src/vmm/src/device_manager/mmio.rs:174:24
>     |
> 174 |           let int_evt = &serial
>     |  ________________________^
> 175 | |             .lock()
> 176 | |             .expect("Poisoned legacy serial lock")
>     | |__________________________________________________^ creates a temporary which is freed while still in use
> 177 |               .interrupt_evt();
>     |                               - temporary value is freed at the end of this statement
> 178 |           vm.register_irqfd(int_evt, self.irq)
>     |                             ------- borrow later used here
>     |
>     = note: consider using a `let` binding to create a longer lived value
> 
> error: aborting due to previous error
> ```
> 
> It's simple to change, but does anyone know why this works on x86_64? This is a lifetime issue so I would have thought not architecture-dependent? Or are optimisers in the different compiler backends re-ordering the code differently?

The function where the error takes place ( i.e. `register_mmio_serial`) is aarch64 labeled, it is not compiled on x86_64.

**Comment by cmannett85** [comment]
Technically these constitute user-facing changes, but they don't change the API, so should there be an entry in `CHANGELOG.md`?

File: src/firecracker/src/main.rs [hunk 1]
```diff
@@ -311,6 +311,6 @@ fn run_without_api(seccomp_filter: BpfProgram, config_json: Option<String>) {
 
     // Run the EventManager that drives everything in the microVM.
     loop {
-        event_manager.run().unwrap();
+        event_manager.run().expect("EventManager exiting");
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/firecracker/src/main.rs]: Done

File: src/polly/src/event_manager.rs [hunk 2]
```diff
@@ -109,7 +109,10 @@ impl EventManager {
     // only used once in this function.
     pub fn add_subscriber(&mut self, subscriber: Arc<Mutex<dyn Subscriber>>) -> Result<()> {
         // Unwrapping here is safe because we want to panic in case the lock is poisoned.
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/polly/src/event_manager.rs]: Done

File: src/polly/src/event_manager.rs [hunk 3]
```diff
@@ -109,7 +109,10 @@ impl EventManager {
...
-        let interest_list = subscriber.lock().unwrap().interest_list();
+        let interest_list = subscriber
+            .lock()
+            .expect("Subscriber lock poisoned")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/polly/src/event_manager.rs]: Done

File: src/polly/src/event_manager.rs [hunk 4]
```diff
@@ -201,10 +204,10 @@ impl EventManager {
             if self.subscribers.contains_key(&pollable) {
                 self.subscribers
                     .get_mut(&pollable)
-                    .unwrap()
+                    .unwrap() // Safe because we have already checked existence
                     .clone()
                     .lock()
-                    .unwrap()
+                    .expect("Poisoned subscriber lock")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/polly/src/event_manager.rs]: Done

File: src/vmm/src/controller.rs [hunk 5]
```diff
@@ -52,7 +52,7 @@ impl VmmController {
     pub fn send_ctrl_alt_del(&mut self) -> ActionResult {
         self.vmm
             .lock()
-            .unwrap()
+            .expect("Poisoned VMM lock")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/vmm/src/controller.rs]: Done

File: src/vmm/src/controller.rs [hunk 6]
```diff
@@ -70,7 +70,7 @@ impl VmmController {
         if let Some(busdev) = self
             .vmm
             .lock()
-            .unwrap()
+            .expect("Poisoned VMM lock")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/vmm/src/controller.rs]: Done

File: src/vmm/src/device_manager/legacy.rs [hunk 7]
```diff
@@ -56,7 +56,7 @@ impl PortIODeviceManager {
         let io_bus = devices::Bus::new();
         let com_evt_1_3 = serial
             .lock()
-            .unwrap()
+            .expect("Poisoned legacy serial lock")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/vmm/src/device_manager/legacy.rs]: Done

File: src/vmm/src/device_manager/mmio.rs [hunk 8]
```diff
@@ -171,8 +171,14 @@ impl MMIODeviceManager {
             return Err(Error::IrqsExhausted);
         }
 
-        vm.register_irqfd(&serial.lock().unwrap().interrupt_evt(), self.irq)
-            .map_err(Error::RegisterIrqFd)?;
+        vm.register_irqfd(
+            &serial
+                .lock()
+                .expect("Poisoned legacy serial lock")
```
[the remark this answers is not in the document]
Comment (cmannett85) [reply src/vmm/src/device_manager/mmio.rs]: Done

File: src/vmm/src/builder.rs [hunk 9]
```diff
@@ -702,7 +704,11 @@ fn attach_net_devices(
         event_manager
             .add_subscriber(net_device.clone())
             .map_err(RegisterEvent)?;
-        let id = net_device.lock().unwrap().id().clone();
+        let id = net_device
+            .lock()
+            .expect("Failed to get net device due to poisoned lock")
```
Comment (acatangiu) [inline src/vmm/src/builder.rs:709]: nit: to save some binary size, I would use a standard common message `Poisoned lock` for all `lock()` fails.
The location of the failing lock is evident from the crash.

Comment (dianpopa) [reply src/vmm/src/builder.rs:709]: I agree with this as long as all messages of this type are replaced.

Comment (cmannett85) [reply src/vmm/src/builder.rs]: Done

**Comment by cmannett85** [comment]
The x86_64 CI tests are failing with this:
```
FAILED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]
...
=================================== FAILURES ===================================
____________________ test_rx_rate_limiting[ubuntu_with_ssh] ____________________
integration_tests/functional/test_rate_limiter.py:152: in test_rx_rate_limiting
    _check_rx_rate_limit_patch(test_microvm, guest_ips)
integration_tests/functional/test_rate_limiter.py:377: in _check_rx_rate_limit_patch
    _check_rx_bandwidth(test_microvm, guest_ips[1], bandwidth_kb)
integration_tests/functional/test_rate_limiter.py:426: in _check_rx_bandwidth
    assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
E   assert 10.228515625 < 10
```
But it is passing on my machine (and aarch64):
```
$ ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting
[Firecracker devtool] 2020-05-06 23:03:15 UTC
[Firecracker devtool] Starting test run ...
================================================================================== test session starts ==================================================================================
platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
cachedir: ../build/pytest_cache
rootdir: /firecracker/tests, inifile: pytest.ini
plugins: timeout-1.3.4
timeout: 300.0s
timeout method: signal
timeout func_only: False
collected 1 item                                                                                                                                                                        

[ unsafe | 1 tests | 1 worker(s) ]
  PASSED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]

================================================================================== 1 passed in 24.85s ===================================================================================
```
I can't see how my changes could have affected this so does anyone have any clues on how to fix it?

**Comment by lauralt** [comment]
> The x86_64 CI tests are failing with this:
> 
> ```
> FAILED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]
> ...
> =================================== FAILURES ===================================
> ____________________ test_rx_rate_limiting[ubuntu_with_ssh] ____________________
> integration_tests/functional/test_rate_limiter.py:152: in test_rx_rate_limiting
>     _check_rx_rate_limit_patch(test_microvm, guest_ips)
> integration_tests/functional/test_rate_limiter.py:377: in _check_rx_rate_limit_patch
>     _check_rx_bandwidth(test_microvm, guest_ips[1], bandwidth_kb)
> integration_tests/functional/test_rate_limiter.py:426: in _check_rx_bandwidth
>     assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
> E   assert 10.228515625 < 10
> ```
> 
> But it is passing on my machine (and aarch64):
> 
> ```
> $ ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting
> [Firecracker devtool] 2020-05-06 23:03:15 UTC
> [Firecracker devtool] Starting test run ...
> ================================================================================== test session starts ==================================================================================
> platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
> cachedir: ../build/pytest_cache
> rootdir: /firecracker/tests, inifile: pytest.ini
> plugins: timeout-1.3.4
> timeout: 300.0s
> timeout method: signal
> timeout func_only: False
> collected 1 item                                                                                                                                                                        
> 
> [ unsafe | 1 tests | 1 worker(s) ]
>   PASSED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]
> 
> ================================================================================== 1 passed in 24.85s ===================================================================================
> ```
> 
> I can't see how my changes could have affected this so does anyone have any clues on how to fix it?


> The x86_64 CI tests are failing with this:
> 
> ```
> FAILED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]
> ...
> =================================== FAILURES ===================================
> ____________________ test_rx_rate_limiting[ubuntu_with_ssh] ____________________
> integration_tests/functional/test_rate_limiter.py:152: in test_rx_rate_limiting
>     _check_rx_rate_limit_patch(test_microvm, guest_ips)
> integration_tests/functional/test_rate_limiter.py:377: in _check_rx_rate_limit_patch
>     _check_rx_bandwidth(test_microvm, guest_ips[1], bandwidth_kb)
> integration_tests/functional/test_rate_limiter.py:426: in _check_rx_bandwidth
>     assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
> E   assert 10.228515625 < 10
> ```
> 
> But it is passing on my machine (and aarch64):
> 
> ```
> $ ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting
> [Firecracker devtool] 2020-05-06 23:03:15 UTC
> [Firecracker devtool] Starting test run ...
> ================================================================================== test session starts ==================================================================================
> platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
> cachedir: ../build/pytest_cache
> rootdir: /firecracker/tests, inifile: pytest.ini
> plugins: timeout-1.3.4
> timeout: 300.0s
> timeout method: signal
> timeout func_only: False
> collected 1 item                                                                                                                                                                        
> 
> [ unsafe | 1 tests | 1 worker(s) ]
>   PASSED    integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting[ubuntu_with_ssh]
> 
> ================================================================================== 1 passed in 24.85s ===================================================================================
> ```
> 
> I can't see how my changes could have affected this so does anyone have any clues on how to fix it?

It is a bug on our side: #1809, sorry for that.

**Comment by cmannett85** [comment]
> Changes look good, please sign your commit: `git commit -s --amend` and update the commit message to follow the [git 50/72 rule](https://preslav.me/2015/02/21/what-s-with-the-50-72-rule/).

Done.

**Comment by acatangiu** [comment]
Great! Please also rebase on top of latest master `git pull upstream master --rebase` and solve conflicts.

**Pull request merged**